File: PlantbreedingPY/generate_lattice_plan.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import random

def generate_lattice_plan(lines, checks, k, n_locs=2, n_reps=2):
    """
    Generate Resolvable Incomplete Block / Alpha Design Plan
    
    Parameters:
    lines (list): List of treatment/line names.
    checks (list): List of check names.
    k (int): Block size (number of plots per incomplete block).
    n_locs (int): Number of locations.
    n_reps (int): Number of full replications per location.
    
    Returns:
    dict: A dictionary containing the randomization DataFrame and the matplotlib Figure object.
    """
    
    treatments = list(lines) + list(checks)
    v = len(treatments) # Total number of treatments
    
    # Validation step: For a resolvable incomplete block design, 
    # total treatments must be a multiple of the incomplete block size.
    if v % k != 0:
        raise ValueError(f"Error: Total treatments ({v}) must be perfectly divisible by incomplete block size k ({k}).")
    
    s = v // k # Number of incomplete blocks per replicate
    
    # ---------------------------------------------------------
    # 1. Randomization Engine
    # ---------------------------------------------------------
    
    plan_list = []
    
    for loc in range(1, n_locs + 1):
        for rep in range(1, n_reps + 1):
            
            # Randomize treatments for the full replication
            shuffled_trt = treatments.copy()
            random.shuffle(shuffled_trt)
            
            # Assign treatments to incomplete blocks
            inc_block_num = np.repeat(np.arange(1, s + 1), k)  # Assign 's' blocks
            local_col = np.tile(np.arange(1, k + 1), s)        # 'k' plots per block
            
            types = ["Check" if t in checks else "Line" for t in shuffled_trt]
            
            temp_df = pd.DataFrame({
                "Location": f"Location {loc}",
                "Replicate": f"Rep {rep}",
                "Rep_Num": rep,
                "Inc_Block_Num": inc_block_num,
                "Local_Col": local_col,
                "Treatment": shuffled_trt,
                "Type": types
            })
            
            # Create a unique identifier for each incomplete block for plotting
            temp_df["Inc_Block"] = f"Rep {rep} - Blk " + temp_df["Inc_Block_Num"].astype(str)
            
            plan_list.append(temp_df)
            
    final_plan = pd.concat(plan_list, ignore_index=True)
    
    # ---------------------------------------------------------
    # 2. Spatial Calculations
    # ---------------------------------------------------------
    
    # Stack Replications vertically. 
    # Each Incomplete Block becomes a row, and the plots within it are columns.
    final_plan["Global_Row"] = final_plan["Inc_Block_Num"] + (final_plan["Rep_Num"] - 1) * s
    final_plan["Global_Col"] = final_plan["Local_Col"]
    
    # ---------------------------------------------------------
    # 3. Plotting
    # ---------------------------------------------------------
    
    locations = final_plan["Location"].unique()
    fig, axes = plt.subplots(1, len(locations), figsize=(6 * len(locations), 8), squeeze=False)
    axes = axes.flatten()
    
    for i, loc_name in enumerate(locations):
        ax = axes[i]
        loc_data = final_plan[final_plan["Location"] == loc_name]
        
        # Draw individual plots (tiles)
        for _, row in loc_data.iterrows():
            color = "#FFD700" if row["Type"] == "Check" else "#ADD8E6"
            rect = patches.Rectangle(
                (row["Global_Col"] - 0.5, row["Global_Row"] - 0.5), 1, 1,
                linewidth=0.5, edgecolor="gray", facecolor=color
            )
            ax.add_patch(rect)
            
            # Add treatment labels
            ax.text(row["Global_Col"], row["Global_Row"], row["Treatment"],
                    ha="center", va="center", fontsize=8, fontweight="bold")
            
        # Draw Incomplete Block boundaries (dashed blue)
        blocks = loc_data.groupby("Inc_Block")
        for _, blk in blocks:
            xmin = blk["Global_Col"].min() - 0.5
            xmax = blk["Global_Col"].max() + 0.5
            ymin = blk["Global_Row"].min() - 0.5
            ymax = blk["Global_Row"].max() + 0.5
            rect = patches.Rectangle(
                (xmin, ymin), xmax - xmin, ymax - ymin,
                linewidth=2, edgecolor="#005b96", facecolor="none", linestyle="dashed"
            )
            ax.add_patch(rect)
            
        # Draw Full Replicate boundaries (bold black)
        reps = loc_data.groupby("Replicate")
        for _, rep_data in reps:
            xmin = rep_data["Global_Col"].min() - 0.5
            xmax = rep_data["Global_Col"].max() + 0.5
            ymin = rep_data["Global_Row"].min() - 0.5
            ymax = rep_data["Global_Row"].max() + 0.5
            rect = patches.Rectangle(
                (xmin, ymin), xmax - xmin, ymax - ymin,
                linewidth=3, edgecolor="black", facecolor="none"
            )
            ax.add_patch(rect)
            
        # Formatting the axes
        ax.set_title(loc_name, fontweight="bold")
        ax.set_xlim(loc_data["Global_Col"].min() - 1, loc_data["Global_Col"].max() + 1)
        ax.set_ylim(loc_data["Global_Row"].min() - 1, loc_data["Global_Row"].max() + 1)
        ax.invert_yaxis() # Match scale_y_reverse()
        ax.axis('off')    # Match theme_minimal() + axis.text.blank()
        
    # Custom Legend
    check_patch = patches.Patch(color='#FFD700', label='Check')
    line_patch = patches.Patch(color='#ADD8E6', label='Line')
    fig.legend(handles=[check_patch, line_patch], loc='lower center', ncol=2, title="Entry Type")
    
    # Title and subtitles
    fig.suptitle("Multi-Location Alpha / Incomplete Block Layout\nBlack borders = Full Replications. Dashed Blue borders = Incomplete Blocks.", 
                 fontweight="bold", y=0.95)
    
    plt.tight_layout()
    fig.subplots_adjust(bottom=0.10, top=0.88) 
    
    plt.show()
    
    return {"Data": final_plan, "Plot": fig}


### ==========================================
### EXAMPLE USAGE
### ==========================================
##
### 24 Total Treatments
##my_lines = [f"L-{i}" for i in range(1, 23)]
##my_checks = ["CHK-1", "CHK-2"]
##
### Run the lattice generator
### We have 24 treatments. We can use an incomplete block size (k) of 4.
### This results in s = 6 incomplete blocks per full replication.
##lattice_trial = generate_lattice_plan(
##    lines = my_lines, 
##    checks = my_checks, 
##    n_locs = 2,              # 2 environments
##    n_reps = 2,              # 2 full replications per environment
##    k = 4                    # 4 plots per incomplete block
##)
##
### You can access the DataFrame using:
### print(lattice_trial["Data"].head())

File: core_logic/r_package/plantbreeding/R/generate_lattice_plan.R
library(dplyr)
library(ggplot2)

#' Generate Resolvable Incomplete Block / Alpha Design Plan
#'
#' @param lines Vector of treatment/line names.
#' @param checks Vector of check names.
#' @param n_locs Number of locations.
#' @param n_reps Number of full replications per location.
#' @param k Block size (number of plots per incomplete block).
#'
#' @return A list containing the randomization data frame and the ggplot object.
generate_lattice_plan <- function(lines, checks, n_locs = 2, n_reps = 2, k) {
  
  treatments <- c(lines, checks)
  v <- length(treatments) # Total number of treatments
  
  # Validation step: For a resolvable incomplete block design, 
  # total treatments must be a multiple of the incomplete block size.
  if (v %% k != 0) {
    stop(paste("Error: Total treatments (", v, 
               ") must be perfectly divisible by incomplete block size k (", k, ")."))
  }
  
  s <- v / k # Number of incomplete blocks per replicate
  
  # ---------------------------------------------------------
  # 1. Randomization Engine
  # ---------------------------------------------------------
  
  plan_list <- list()
  counter <- 1
  
  for (loc in 1:n_locs) {
    for (rep in 1:n_reps) {
      
      # Randomize treatments for the full replication
      shuffled_trt <- sample(treatments)
      
      # Assign treatments to incomplete blocks
      temp_df <- data.frame(
        Location = paste("Location", loc),
        Replicate = paste("Rep", rep),
        Rep_Num = rep,
        Inc_Block_Num = rep(1:s, each = k),  # Assign 's' blocks
        Local_Col = rep(1:k, times = s),     # 'k' plots per block
        Treatment = shuffled_trt,
        Type = ifelse(shuffled_trt %in% checks, "Check", "Line")
      )
      
      # Create a unique identifier for each incomplete block for plotting
      temp_df$Inc_Block <- paste("Rep", rep, "- Blk", temp_df$Inc_Block_Num)
      
      plan_list[[counter]] <- temp_df
      counter <- counter + 1
    }
  }
  
  final_plan <- bind_rows(plan_list)
  
  # ---------------------------------------------------------
  # 2. Spatial Calculations
  # ---------------------------------------------------------
  
  # Stack Replications vertically. 
  # Each Incomplete Block becomes a row, and the plots within it are columns.
  final_plan <- final_plan %>%
    group_by(Location) %>%
    mutate(
      Global_Row = Inc_Block_Num + (Rep_Num - 1) * s,
      Global_Col = Local_Col
    ) %>%
    ungroup()
  
  # Bounding boxes for Full Replicates (Bold Black)
  rep_boundaries <- final_plan %>%
    group_by(Location, Replicate) %>%
    summarize(
      xmin = min(Global_Col) - 0.5,
      xmax = max(Global_Col) + 0.5,
      ymin = min(Global_Row) - 0.5,
      ymax = max(Global_Row) + 0.5,
      .groups = "drop"
    )
  
  # Bounding boxes for Incomplete Blocks (Dashed Blue)
  blk_boundaries <- final_plan %>%
    group_by(Location, Inc_Block) %>%
    summarize(
      xmin = min(Global_Col) - 0.5,
      xmax = max(Global_Col) + 0.5,
      ymin = min(Global_Row) - 0.5,
      ymax = max(Global_Row) + 0.5,
      .groups = "drop"
    )
  
  # ---------------------------------------------------------
  # 3. Plotting
  # ---------------------------------------------------------
  
  field_plot <- ggplot() +
    # Draw individual plots (thin gray lines)
    geom_tile(data = final_plan, 
              aes(x = Global_Col, y = Global_Row, fill = Type), 
              color = "gray80", linewidth = 0.3) +
    
    # Draw Incomplete Block boundaries (dashed blue)
    geom_rect(data = blk_boundaries, 
              aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), 
              fill = NA, color = "#005b96", linewidth = 0.8, linetype = "dashed") +
    
    # Draw Full Replicate boundaries (bold black)
    geom_rect(data = rep_boundaries, 
              aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax), 
              fill = NA, color = "black", linewidth = 1.2) +
    
    # Add treatment labels
    geom_text(data = final_plan, 
              aes(x = Global_Col, y = Global_Row, label = Treatment), 
              size = 3.5, fontface = "bold") +
    
    scale_fill_manual(values = c("Check" = "#FFD700", "Line" = "#ADD8E6")) +
    scale_y_reverse() + 
    scale_x_continuous() +
    facet_wrap(~ Location, ncol = 2, scales = "free") + 
    labs(title = "Multi-Location Alpha / Incomplete Block Layout",
         subtitle = "Black borders = Full Replications. Dashed Blue borders = Incomplete Blocks.",
         x = "Field Column (Plot within Block)",
         y = "Field Row (Incomplete Block)",
         fill = "Entry Type") +
    theme_minimal() +
    theme(
      panel.grid = element_blank(),
      strip.text = element_text(size = 12, face = "bold"),
      axis.text = element_blank() 
    )
  
  print(field_plot)
  return(list(Data = final_plan, Plot = field_plot))
}

# ==========================================
# EXAMPLE USAGE
# ==========================================

# 24 Total Treatments
my_lines <- paste0("L-", 1:22)
my_checks <- c("CHK-1", "CHK-2")

# Run the lattice generator
# We have 24 treatments. We can use an incomplete block size (k) of 4.
# This results in s = 6 incomplete blocks per full replication.
lattice_trial <- generate_lattice_plan(
  lines = my_lines, 
  checks = my_checks, 
  n_locs = 2,             # 2 environments
  n_reps = 2,             # 2 full replications per environment
  k = 4                   # 4 plots per incomplete block
)